File: rust/src/models/customer_details.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{address::Address, customer_id::CustomerId};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct CustomerDetails {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alias: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub billing_address: Option<Address>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,

    pub id: CustomerId,

    pub name: String,

    pub snapshot_at: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub vat_number: Option<String>,
}

impl CustomerDetails {
    pub fn new(id: CustomerId, name: String, snapshot_at: String) -> Self {
        Self {
            alias: None,
            billing_address: None,
            email: None,
            id,
            name,
            snapshot_at,
            vat_number: None,
        }
    }
}

File: typescript/src/models/customerDetails.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type Address, AddressSerializer } from "./address";
import { type CustomerId, CustomerIdSerializer } from "./customerId";

export interface CustomerDetails {
  alias?: string | null;

  billingAddress?: Address | null;

  email?: string | null;

  id: CustomerId;

  name: string;

  snapshotAt: Date;

  vatNumber?: string | null;
}

export const CustomerDetailsSerializer = {
  _fromJsonObject(object: any): CustomerDetails {
    return {
      alias: object["alias"],
      billingAddress:
        object["billing_address"] != null
          ? AddressSerializer._fromJsonObject(object["billing_address"])
          : undefined,
      email: object["email"],
      id: CustomerIdSerializer._fromJsonObject(object["id"]),
      name: object["name"],
      snapshotAt: parseDateTime(object["snapshot_at"]),
      vatNumber: object["vat_number"],
    };
  },

  _toJsonObject(self: CustomerDetails): any {
    return {
      alias: self.alias,
      billing_address:
        self.billingAddress != null
          ? AddressSerializer._toJsonObject(self.billingAddress)
          : undefined,
      email: self.email,
      id: CustomerIdSerializer._toJsonObject(self.id),
      name: self.name,
      snapshot_at: self.snapshotAt,
      vat_number: self.vatNumber,
    };
  },
};
